### PR TITLE
[Resolver] Take supported platforms into account when resolving

### PR DESCRIPTION
**Pull request by segiddins**
Closes https://github.com/CocoaPods/CocoaPods/issues/2443.
The code style is terrible, but I want some feedback before I try and clean this up.

### CODE REVIEW

**Comment by orta** [comment]
Is this in preparation for more platform support?

**Comment by segiddins** [comment]
@orta no, it's just addressing a long-standing feature request

**Comment by neonichu** [comment]
Generally looks reasonable to me.

File: lib/cocoapods/resolver.rb [hunk 1]
```diff
@@ -395,6 +400,17 @@ def handle_resolver_error(error)
       raise Informative, error.message
     end
 
+    def plat?(dg, req, spec)
```
Comment (kylef) [inline lib/cocoapods/resolver.rb]: Can we use some more explanatory variable names to make this readable to people who didn't write the resolver/this particular piece of code? Ex: `dependency_graph` not `dg`

Comment (segiddins) [inline lib/cocoapods/resolver.rb]: Please read the PR description:
> The code style is terrible, but I want some feedback before I try and clean this up.

**Comment by segiddins** [comment]
This is ready for review. I'd also like to ask @alloy to take a look at this?

**Comment by neonichu** [comment]
LGTM :+1:

File: lib/cocoapods/resolver.rb [hunk 2]
```diff
@@ -52,7 +53,11 @@ def initialize(sandbox, podfile, locked_dependencies, sources)
     #         definition.
     #
     def resolve
-      dependencies = podfile.target_definition_list.map(&:dependencies).flatten
+      dependencies = podfile.target_definition_list.flat_map do |target|
+        target.dependencies.each do |dep|
+          @platforms_by_dependency[dep].<<(target.platform).uniq!
```
Comment (alloy) [inline lib/cocoapods/resolver.rb]: Hah, maybe use `#push` instead of `#<<` here?

File: lib/cocoapods/resolver.rb [hunk 3]
```diff
@@ -395,6 +400,27 @@ def handle_resolver_error(error)
...
+    # @return [Bool]
+    def spec_is_platform_compatible?(dependency_graph, dependency, spec)
+      all_predecessors = ->(vertex) do
+        pred = vertex.predecessors
+        pred + pred.map(&all_predecessors).reduce(Set.new, &:&) << vertex
+      end
+      vertex = dependency_graph.vertex_named(dependency.name)
+      predecessors = all_predecessors[vertex].reject { |v| v.explicit_requirements.empty? }
+      platforms_to_satisfy = predecessors.flat_map(&:explicit_requirements).flat_map { |r| @platforms_by_dependency[r] }
+      platforms_to_satisfy.all? { |pts| spec.available_platforms.select { |p| p.name == pts.name }.all? { |p| pts.supports?(p) } }
```
Comment (alloy) [inline lib/cocoapods/resolver.rb]: I’d personally like to see this spread over a few more lines and I’m not a fan of too many short variable names.

**Comment by alloy** [comment]
Logic looks good to me :+1: Just some stylistic points.

**Pull request merged**